### PR TITLE
Feature/authentication using jwt token

### PR DESCRIPTION
**Pull request by chitranshubisht**
```
Added specs for users_controller and posts_controller
Implemented JWT authorization for login
```

### CODE REVIEW

File: .byebug_history [hunk 1]
```diff
@@ -0,0 +1,18 @@
+exit
+@user.authenticate("123")
+@user.authenticate({"123"})
+@user.authenticate("123")
+@user.password_digest
+@user.password
+@user
+@user&.authenticate(params[:password_digest])
+params[:password_digest]
+params
```
Comment (nyadav-deqode) [inline .byebug_history]: Why did you push byebug history

Comment (nyadav-deqode) [reply .byebug_history]: Please update your branch with master

**Review by nyadav-deqode — CHANGES_REQUESTED** [review]

File: .byebug_history (hunk 1 above)
Comment (chitranshubisht) [reply .byebug_history]: Removed byebug history file.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -13,9 +15,25 @@ def index
...
+
   private
 
   def permitted_params
-    params.require(:user).permit(:name, :email)
+    params.require(:user).permit(:name, :email, :password)
+  end
+
+  def find_user
+    @user = User.find(params[:id])
```
Comment (nyadav-deqode) [inline app/controllers/users_controller.rb:38]: This will raise an error for other actions like index, create

**Review by nyadav-deqode — CHANGES_REQUESTED** [review]

File: app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,2 +1,18 @@
...
+  before_action :authenticate_user
+
+  def authenticate_user
+    header = request.headers['Authorization']
+    header = header.split(' ').last if header
+    begin
+      @decoded = jwt_decode(header)
+      @current_user = User.find(@decoded[:user_id])
+    rescue ActiveRecord::RecordNotFound => e
+      render json: { errors: e.message }, status: :unauthorized
```
Comment (nyadav-deqode) [inline app/controllers/application_controller.rb]: Avoid repetition of rescue, make it DRY

File: app/controllers/posts_controller.rb [hunk 4]
```diff
@@ -1,4 +1,5 @@
 class PostsController < ApplicationController
+  include JwtToken
```
Comment (nyadav-deqode) [inline app/controllers/posts_controller.rb]: This include is repeated in all the controllers, avoid repetition

**Pull request merged**